File: cip/src/formats/pdf_parser.cpp
// Copyright (c) 2022 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by GNU General Public License
// that can be found in the LICENSE file.

#include "formats/pdf_parser.h"

#include <QDebug>

#include "formats/pdf_reader.h"
#include "formats/util.h"

bool ParsePdfFile(const QString& filepath) {
  PdfReader reader;
  if (!reader.load(filepath)) {
    qWarning() << "Failed to open pdf file:" << filepath;
    return false;
  }
  const int pages = reader.numPages();
  QString text;
  bool ok;

  // First 10 pages.
  int front_page = 0;
  for (front_page = 0; front_page < 10 && front_page < pages; ++front_page) {
    ok = reader.readPage(front_page, text);
    if (ok && IsPlainCipPage(text)) {
      return ParsePdfMetadata(filepath, text);
    }
  }

  // Last 5 pages.
  for (int back_page = qMax(pages - 5, front_page); back_page < pages; ++back_page) {
    ok = reader.readPage(back_page, text);
    if (ok && IsPlainCipPage(text)) {
      return ParsePdfMetadata(filepath, text);
    }
  }

  qWarning() << "No cip page found in:" << filepath;
  return false;
}

bool ParsePdfMetadata(const QString& filepath, const QString& text) {
  Q_UNUSED(filepath);
  Q_UNUSED(text);
//  qDebug() << qPrintable(text);
  return true;
}

File: cip/src/parsers/pdf_parser.rs
// Copyright (c) 2023 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by GNU General Public License
// that can be found in the LICENSE file.

use std::path::Path;

use crate::error::{Error, ErrorKind};
use crate::formats::pdf_reader::PdfReader;
use crate::parsers::cip::{is_plain_cip_page, parse_cip_from_text, CipRecord};

pub fn parse_cip_from_pdf<P: AsRef<Path>>(path: P) -> Result<CipRecord, Error> {
    let reader = PdfReader::open(path)?;
    let pages = reader.pages();

    // First 10 pages.
    let mut front_page = 1;
    while front_page < 10 && front_page < pages {
        let text = reader.read_page(front_page)?;
        println!("page: {front_page}, text: {text}");
        if is_plain_cip_page(&text) {
            return parse_cip_from_text(&text);
        }

        front_page += 1;
    }

    // Last 5 pages.
    let mut rear_page = (pages - 5).max(front_page);
    while rear_page < pages {
        let text = reader.read_page(rear_page)?;
        println!("rear_page: {rear_page}, text: {text}");
        if is_plain_cip_page(&text) {
            return parse_cip_from_text(&text);
        }

        rear_page += 1;
    }

    Err(Error::new(
        ErrorKind::NoCipRecordFound,
        "No cip record found in pdf file",
    ))
}
